### PR TITLE
Add support for set api Modals

### PR DESCRIPTION
**Pull request by Kowlin**
### Description of the changes
* Changes ``set api`` to have the default options to spawn as a modal
* Allowed developers to call an ``SetApiView`` and ``SetApiModal`` to set API keys in a secure manner.
* These modals and views are hard coded to only accept bot owners.

### Have the changes in this PR been tested?
Yes

### CODE REVIEW

File: redbot/core/utils/views.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        self.default_service = default_service
+        self.default_keys: List[str] = []
+        if default_keys is not None:
+            self.default_keys = list(default_keys.keys())
+        self.default_keys_fmt = self._format_keys(default_keys)
+
+        _placeholder_service = "service"
+        if self.default_service is not None:
+            _placeholder_service = self.default_service
+        _placeholder_token = "client_id YOUR_CLIENT_ID"
```
Comment (MeatyChunks) [inline redbot/core/utils/views.py]: Maybe client_secret can go here too? Clearing up any confusion of there being only one modal to fill in.

File: redbot/core/utils/views.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+                _("This button is for bot owners only. Oh well"), ephemeral=True
+            )
+            return False
+        return True
+
+    @discord.ui.button(
+        label=_("Set API token"),
+        style=discord.ButtonStyle.grey,
+    )
+    async def auth_button(self, button: discord.Button, interaction: discord.Interaction):
```
Comment (MeatyChunks) [inline redbot/core/utils/views.py]: Just your luck - interaction parameter order changed requiring the interaction before the button now, lol.

File: redbot/core/utils/views.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        Defaults to ``None``.
+    default_keys: Optional[Dict[str, str]]
+        The API keys the service is expecting.
+        This will only allow the bot owner to set keys the Modal is expecting.
+        Defaults to ``None``.
+    """
+
+    def __init__(
+        self,
+        bot: Red,
```
Comment (TrustyJAID) [inline redbot/core/utils/views.py]: Turns out, you don't need to pass the bot into this. `interaction.client` is the bot.

File: redbot/core/utils/views.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+                else:
+                    ret += f"{k} YOUR_{k.upper()}\n"
+            return ret
+        else:
+            return None
+
+    async def on_submit(self, interaction: discord.Interaction):
+        if (
+            await self.bot.is_owner(interaction.user) is False
+        ):  # Prevent non-bot owners from somehow aquiring and saving the modal.
```
Comment (TrustyJAID) [inline redbot/core/utils/views.py]:
```suggestion
        if (
            not await interaction.client.is_owner(interaction.user)
        ):  # Prevent non-bot owners from somehow aquiring and saving the modal.
```

File: redbot/core/utils/views.py (hunk 3 above)
Comment (TrustyJAID) [inline redbot/core/utils/views.py]:
```suggestion
```

File: redbot/core/utils/views.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        try:
+            tokens = await converter().convert(None, tokens)
+        except BadArgument as exc:
+            return await interaction.response.send_message(
+                _("{error_message}\nPlease try again.").format(error_message=str(exc)),
+                ephemeral=True,
+            )
+
+        if self.default_service is not None:  # Check is there is a service set.
+            await self.bot.set_shared_api_tokens(self.default_service, **tokens)
```
Comment (TrustyJAID) [inline redbot/core/utils/views.py]:
```suggestion
            await interaction.client.set_shared_api_tokens(self.default_service, **tokens)
```

File: redbot/core/utils/views.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+            return await interaction.response.send_message(
+                _("`{service}` API tokens have been set.").format(service=self.default_service),
+                ephemeral=True,
+            )
+        else:
+            service = self.service_input.value.lower()
+            await self.bot.set_shared_api_tokens(service, **tokens)
```
Comment (TrustyJAID) [inline redbot/core/utils/views.py]:
```suggestion
            await interaction.client.set_shared_api_tokens(service, **tokens)
```

File: redbot/core/utils/views.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+        Defaults to ``None``.
+    default_keys: Optional[Dict[str, str]]
+        The API keys the service is expecting.
+        This will only allow the bot owner to set keys the Modal is expecting.
+        Defaults to ``None``.
+    """
+
+    def __init__(
+        self,
+        bot: Red,
```
Comment (TrustyJAID) [inline redbot/core/utils/views.py]:
```suggestion
```

File: redbot/core/utils/views.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+        default_service: Optional[str] = None,
+        default_keys: Optional[Union[str, List[str]]] = None,
+    ):
+        self.bot: Red = bot
```
Comment (TrustyJAID) [inline redbot/core/utils/views.py]:
```suggestion
```

File: redbot/core/utils/views.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+        self.default_service = default_service
+        self.default_keys = default_keys
+        super().__init__()
+
+    async def interaction_check(self, interaction: discord.Interaction) -> bool:
+        if await self.bot.is_owner(interaction.user) is False:
```
Comment (TrustyJAID) [inline redbot/core/utils/views.py]:
```suggestion
        if not await interaction.client.is_owner(interaction.user):
```

File: redbot/core/utils/views.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+        Defaults to ``None``.
+    """
+
+    def __init__(
+        self,
+        bot: Red,
+        default_service: Optional[str] = None,
+        default_keys: Optional[Dict[str, str]] = None,
+    ):
+        self.bot: Red = bot
```
Comment (TrustyJAID) [inline redbot/core/utils/views.py]:
```suggestion
```

File: redbot/core/core_commands.py [hunk 11]
```diff
@@ -3091,29 +3092,49 @@ async def _set_regional_format_local(self, ctx: commands.Context, language_code:
...
-        await ctx.bot.set_shared_api_tokens(service, **tokens)
-        await ctx.send(_("`{service}` API tokens have been set.").format(service=service))
+        if service is None:  # Handled in order of missing operations
+            await ctx.send(
+                _("Click the button below to set your keys."), view=SetApiView(self.bot)
+            )
+        elif tokens is None:
+            await ctx.send(
+                _("Click the button below to set your keys."),
+                view=SetApiView(self.bot, default_service=service),
```
Comment (TrustyJAID) [inline redbot/core/core_commands.py]:
```suggestion
                view=SetApiView(default_service=service),
```

File: redbot/core/utils/views.py [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+
+import discord
+
+from discord.ext.commands import BadArgument
+from typing import TYPE_CHECKING, List, Dict, Union, Optional
+from redbot.core.commands.converter import get_dict_converter
+from redbot.core.i18n import Translator
+
+if TYPE_CHECKING:
+    from redbot.core.bot import Red
```
Comment (TrustyJAID) [inline redbot/core/utils/views.py]:
```suggestion
```

File: redbot/core/utils/views.py [hunk 13]
```diff
@@ -0,0 +1,187 @@
+from __future__ import annotations
+
+import discord
+
+from discord.ext.commands import BadArgument
+from typing import TYPE_CHECKING, List, Dict, Union, Optional
```
Comment (TrustyJAID) [inline redbot/core/utils/views.py]:
```suggestion
from typing import List, Dict, Union, Optional
```

**Review by TrustyJAID — CHANGES_REQUESTED** [review]

File: redbot/core/utils/views.py [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+    default_keys: Optional[Dict[str, str]]
+        The API keys the service is expecting.
+        This will only allow the bot owner to set keys the Modal is expecting.
+        Defaults to ``None``.
+    """
+
+    def __init__(
+        self,
+        default_service: Optional[str] = None,
+        default_keys: Optional[Union[str, List[str]]] = None,
```
Comment (TrustyJAID) [inline redbot/core/utils/views.py]:
```suggestion
        default_keys: Optional[Dict[str, str]] = None,
```

**Review by TrustyJAID — CHANGES_REQUESTED** [review]
One minor change

File: redbot/core/core_commands.py [hunk 15]
```diff
@@ -18,7 +18,8 @@
 import traceback
 from pathlib import Path
 from redbot.core import data_manager
-from redbot.core.utils.menus import menu
+from redbot.core.utils.menus import menu, DEFAULT_CONTROLS
```
Comment (TrustyJAID) [inline redbot/core/core_commands.py]: Also DEFAULT_CONTROLS is not imported in master anymore
```suggestion
from redbot.core.utils.menus import menu
```

**Review by TrustyJAID — APPROVED** [review]
Looks good to me!

**Pull request merged**